src/analyses/base.ml

```diff
@@ -2430,6 +2437,20 @@ struct
     | Events.AssignSpawnedThread (lval, tid) ->
       (* TODO: is this type right? *)
       set ~ctx:(Some ctx) (Analyses.ask_of_ctx ctx) ctx.global ctx.local (eval_lv (Analyses.ask_of_ctx ctx) ctx.global ctx.local lval) (Cilfacade.typeOfLval lval) (`Thread (ValueDomain.Threads.singleton tid))
+    | Events.Invalidate {addr_list} ->
+      let invalidate_address a =
+        let t = AD.get_type a in
+        let v = get ask ctx.global st a None in (* None here is ok, just causes us to be a bit less precise *)
+        let nv =  VD.invalidate_value ask t v in
+        (a, t, nv)
+      in
+      let invalids = List.map invalidate_address addr_list in
+      if M.tracing && addr_list <> [] then (
+        let addrs = List.map (Tuple3.first) invalids in
+        let vs = List.map (Tuple3.third) invalids in
+        M.tracel "invalidate" "Setting addresses [%a] to values [%a]\n" (d_list ", " AD.pretty) addrs (d_list ", " VD.pretty) vs
+      );
+      set_many ~ctx ask ctx.global ctx.local invalids
```

Comment:
What's the reason for converting this into an event instead of simply extracting the functions for all globals invalidation etc inside the base analysis itself?
All the calls to `invalidate_core` look awkward, because they have to delegate outside to the events system to perform these after the actual transfer function completes, but in all three cases the transfer function just returns `ctx.local` as the updated state. This seems like an unnecessarily roundabout way to extract common logic.

Reply:
>What's the reason for converting this into an event instead of simply extracting the functions for all globals invalidation etc inside the base analysis itself?

The idea of making invalidation an event based mechanism was that other analyses could be informed and then do some proper handling. One could e.g. imagine that an relational analysis that handles referenced variables might need to know when some variables are potentially invalidated.

Reply:
I'm just afraid that in light of #719 with the need to distinguish non-transitive vs transitive accesses also for reads/frees and writes other than invalidation, this won't be general enough.

It's also quite base analysis specific by using its `ValueDomain.Addr`. Other analyses like apron or var_eq, which currently also do similar invalidation, aren't using base analysis's address abstraction, but have their own ways of working with invalidated lvalues/rvalues.

Reply:
>I'm just afraid that in light of https://github.com/goblint/analyzer/issues/719 with the need to distinguish non-transitive vs transitive accesses also for reads/frees and writes other than invalidation, this won't be general enough.

The idea for invalidate events would be that the `Base` analysis collects all lvals that have to be invalidated. This way, other analyses would not have to implement the collection of transitively reachable lvals.

>It's also quite base analysis specific by using its ValueDomain.Addr

One could add a utility function that takes elements from the `ValueDomain.Addr` and translates them back to the `lval` they are on (without offset information, in case the offset is not concrete).

Reply:
> The idea for invalidate events would be that the `Base` analysis collects all lvals that have to be invalidated. This way, other analyses would not have to implement the collection of transitively reachable lvals.

They don't currently reimplement it either, but just use the `ReachableFrom` query where needed. Since #590 those queries are cached, so multiple analyses asking for the same reachability comes at no extra cost.

We will probably have to see how this plays together with the general non-transitive vs transitive accessing changes from #719 and #720.

Reply:
I thought about this more and how it compares to #720 and the `Access` event. It might make sense to have the `Invalidate` event after all because:
1. `Access` event is emitted for all writes, not just invalidating ones. Although this could be changed by adding an extra field to that event.
2. `Access` event is emitted only during multithreaded mode, because it's all about race detection. But these invalidations also need to happen in singlethreaded mode. Although this could also be changed (and accesses from singlethreaded mode wouldn't race with anything).

But given that access analysis (and its event) is really about race detection, it might be better to keep it that way.